src/lib/hash/mdx_hash/mdx_hash.h

```diff
@@ -1,62 +1,120 @@
 /*
-* MDx Hash Function
-* (C) 1999-2008 Jack Lloyd
+* (C) 1999-2008,2023 Jack Lloyd
 *
 * Botan is released under the Simplified BSD License (see license.txt)
 */
 
-#ifndef BOTAN_MDX_BASE_H_
-#define BOTAN_MDX_BASE_H_
+#ifndef BOTAN_MDX_HELPER_H_
+#define BOTAN_MDX_HELPER_H_
 
-#include <botan/hash.h>
+#include <botan/internal/bit_ops.h>
+#include <botan/internal/loadstor.h>
 
 namespace Botan {
 
-/**
-* MDx Hash Function Base Class
-*/
-class MDx_HashFunction : public HashFunction
+enum class MD_Endian {
+   Little,
+   Big,
+};
+
+template<MD_Endian ENDIAN,
+         typename DIGEST_T,
+         size_t DIGEST_ELEM,
+         void init_fn(DIGEST_T[DIGEST_ELEM]),
+         void compress_fn(DIGEST_T[DIGEST_ELEM], const uint8_t[], size_t),
+         size_t BLOCK_BYTES = 64,
+         size_t DIGEST_LENGTH = DIGEST_ELEM * sizeof(DIGEST_T),
+         size_t CTR_BYTES = 8>
+class MD_Hash final
```

Comment:
Let me suggest an alternative design using C++20 concepts and a CRTP-ish structure: Instead of this long list of parameters, we could write a concept to tie down the parameters each MDx-hash needs to define (as static constexpr members). Then, the instantiation of `MDx_Hash` in the concrete implementations shrinks to: `MDx_Hash<MD5> m_md;`:

```C++
template <typename T>
concept mdx_hash_implementation =
    requires(T::digest_type digest, uint8_t input[], size_t blocks, MD_Endian endian) {
        typename T::digest_type;
        T::endianness;
        T::block_bytes;
        T::ctr_bytes;
        T::init(digest);
        T::compress_n(digest, input, blocks);
    } &&
    T::block_bytes >= 64 && is_power_of_2(T::block_bytes) &&
    T::ctr_bytes >= 8 && is_power_of_2(T::ctr_bytes) &&
    T::ctr_bytes < T::block_bytes &&
    sizeof(typename T::digest_type) >= 16;
    

template <mdx_hash_implementation T>
class MDx
{
    std::array<uint8_t, T::block_bytes> m_buffer;
    T::digest_type m_digest;
    uint64_t m_count;
    size_t m_position;
};

class MD5
{
public:
    using digest_type = std::array<uint32_t, 4>;
    static constexpr MD_Endian endianness = MD_Endian::Little;
    static constexpr size_t block_bytes = 64;
    static constexpr size_t ctr_bytes = 8;
    
public:
    static void init(digest_type& digest);
    static void compress_n(digest_type& digest, const uint8_t input[], size_t blocks);

private:
    MDx<MD5> m_md;
};
```

That surely is a bit more boilerplate in the concrete implementations. But I see a few advantages:

* It lends itself from the usual STL members: e.g. `std::vector<>::value_type`
* Readability: Instead of a long list of magic numbers in the `MDx_Hash` instantiation, each parameter has a speaking name
* DRY: Parameter definitions can be referred to in the implementation classes (if needed) -> less implicit knowledge: e.g. `digest_type` encapsulates the `digest_length` and `digest_element_type` and the `init()`/`compress_n()` methods don't need to repeat it
* Parameter constraints are declared as a concept, hopefully resulting in better compiler errors

Certainly this can be further refined.

Godbolt: https://godbolt.org/z/1h1nzcYoh